docs/helper-environment-visitor.md

```diff
@@ -0,0 +1,285 @@
+
+```js title="my-babel-plugin.js"
+if (path.isMethod()) {
+  requeueComputedKeyAndDecorators(path)
+}
+```
+
+## Example
+
+### Replace top level `this`
+
+Suppose we are migrating vanilla JavaScript to ES Modules. Now that the `this` keyword is equivalent to `undefined` at the top level of an ESModule ([spec](https://tc39.es/ecma262/#sec-module-environment-records-getthisbinding)), we want to replace all top-level `this` to [`globalThis`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis):
+
+```js title=input.js
+// replace this expression to `globalThis.foo = "top"`
+this.foo = "top";
+
+() => {
+  // replace
+  this.foo = "top"
+}
+```
+
+We can draft a code mod plugin, here is our first revision:
+
+```js title="Revision 1: replace-top-level-this-plugin.js"
+module.exports = (api) => {
+  const { types: t } = api;
+  return {
+    name: "replace-top-level-this",
+    visitor: {
+      ThisExpression(path) {
+        path.replaceWith(t.identifier("globalThis"));
+      }
+    }
+  }
+}
+```
+
+The first revision works for examples so far. However, it does not really capture the idea of top-level: For example, we should not replace `this` within a non-arrow function: e.g. function declaration, object methods and class methods:
+
+```js title="input.js"
+function Foo() {
+  // don't replace
+  this.foo = "inner";
+}
+
+class Bar {
+  method() {
+    // don't replace
+    this.foo = "inner";
+  }
+}
+```
+
+We can skip traversing if we encounter such non-arrow functions. Here we combine multiple AST types with `|` in the visitor selector.
+
+```js title="Revision 2: replace-top-level-this-plugin.js"
+module.exports = (api) => {
+  const { types: t } = api;
+  return {
+    name: "replace-top-level-this",
+    visitor: {
+      ThisExpression(path) {
+        path.replaceWith(t.identifier("globalThis"));
+      }
+      // highlight-start
+      "FunctionDeclaration|FunctionExpression|ObjectMethod|ClassMethod|ClassPrivateMethod"(path) {
+        path.skip();
+      }
+      // highlight-end
+    }
+  }
+}
+```
+
+`"FunctionDeclaration|..."` is a really long string, so it is not maintainable. We can
+improve that using the [FunctionParent](types.md#functionparent) alias:
+
+```js title="Revision 3: replace-top-level-this-plugin.js"
+module.exports = (api) => {
+  const { types: t } = api;
+  return {
+    name: "replace-top-level-this",
+    visitor: {
+      ThisExpression(path) {
+        path.replaceWith(t.identifier("globalThis"));
+      }
+      // highlight-start
+      FunctionParent(path) {
+        if (!path.isArrowFunctionExpression()) {
+          path.skip();
+        }
+      }
+      // highlight-end
+    }
+  }
+}
+```
+
+The plugin works generally. However, it can not handle an edge case where top-level `this` is in computed class elements:
```

Comment:
```suggestion
The plugin works generally. However, it can not handle an edge case where top-level `this` is used within computed class elements:
```